pandas/core/reshape/merge.py

```diff
@@ -915,11 +916,15 @@ def _maybe_coerce_merge_keys(self):
             # Houston, we have a problem!
             # let's coerce to object
             if name in self.left.columns:
+                cat = is_categorical_dtype(lk)
+                typ = lk.categories.dtype if cat else object
```

Comment:
I think you can do this above L898, e.g. something like

```
lk_to = object
rk_to = object
....
# L898
           elif is_categorical_dtype(lk) or is_categorical_dtype(rk):
               if is_categorical_dtype(lk):
                    lk_to = lk.categories.dtype
               if is_categorycal_dtype(rk):
                    rk_to = rk.categories.dtype
```

then use ``lk_to`` and ``rk_to`` where you used ``typ``